File: sdk/examples/how_tos/account/state_transition.rs
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we will update the state metadata of an account output.
//!
//! Make sure that `STRONGHOLD_SNAPSHOT_PATH` and `WALLET_DB_PATH` already exist by
//! running the `./how_tos/accounts_and_addresses/create_account.rs` example!
//!
//! Rename `.env.example` to `.env` first, then run the command:
//! ```sh
//! cargo run --release --all-features --example state_transition
//! ```

use iota_sdk::{types::block::output::AccountOutputBuilder, wallet::Result, Wallet};

// The metadata for the next state
const NEW_STATE_METADATA: &str = "updated state metadata 1";

#[tokio::main]
async fn main() -> Result<()> {
    //  This example uses secrets in environment variables for simplicity which should not be done in production.
    dotenvy::dotenv().ok();

    let wallet = Wallet::builder()
        .with_alias("Alice")
        .with_storage_path(&std::env::var("WALLET_DB_PATH").unwrap())
        .finish()
        .await?;

    // Set the stronghold password
    wallet
        .set_stronghold_password(std::env::var("STRONGHOLD_PASSWORD").unwrap())
        .await?;

    // May want to ensure the account is synced before sending a transaction.
    let balance = wallet.sync(None).await?;

    // Get the first account
    let account_id = balance
        .accounts()
        .first()
        .expect("No account output available in the account.");

    let account_output_data = wallet
        .unspent_account_output(account_id)
        .await?
        .expect("account not found in unspent outputs");
    println!(
        "Account '{account_id}' found in unspent output: '{}'",
        account_output_data.output_id
    );

    let token_supply = wallet.client().get_token_supply().await?;
    let rent_structure = wallet.client().get_rent_structure().await?;

    let account_output = account_output_data.output.as_account();
    let updated_account_output = AccountOutputBuilder::from(account_output)
        // Minimum required storage deposit will change if the new metadata has a different size, so we will update
        // the amount
        .with_minimum_storage_deposit(rent_structure)
        .with_state_metadata(NEW_STATE_METADATA.as_bytes().to_vec())
        .with_state_index(account_output.state_index() + 1)
        .finish_output(token_supply)?;

    println!("Sending transaction...",);
    let transaction = wallet.send_outputs(vec![updated_account_output], None).await?;
    println!("Transaction sent: {}", transaction.transaction_id);

    let block_id = wallet
        .reissue_transaction_until_included(&transaction.transaction_id, None, None)
        .await?;
    println!(
        "Block included: {}/block/{}",
        std::env::var("EXPLORER_URL").unwrap(),
        block_id
    );

    Ok(())
}

File: bindings/python/examples/how_tos/alias/state_transition.py
import os

from dotenv import load_dotenv

from iota_sdk import Wallet, FilterOptions, utf8_to_hex

load_dotenv()

# In this example we will update the state metadata of an alias output.

NEW_STATE_METADATA = 'updated state metadata 1'

if 'WALLET_DB_PATH' not in os.environ:
    raise Exception(".env WALLET_DB_PATH is undefined, see .env.example")

if 'STRONGHOLD_PASSWORD' not in os.environ:
    raise Exception(".env STRONGHOLD_PASSWORD is undefined, see .env.example")

if 'EXPLORER_URL' not in os.environ:
    raise Exception(".env EXPLORER_URL is undefined, see .env.example")

wallet = Wallet(os.environ['WALLET_DB_PATH'])

account = wallet.get_account('Alice')

# Sync account with the node
balance = account.sync()

if len(balance.aliases) == 0:
    raise Exception("No Alias available in account 'Alice'")

alias_id = balance.aliases[0]

alias_output_data = account.unspent_outputs(
    FilterOptions(aliasIds=[alias_id]))[0]
print(f"Alias {alias_id} found in unspent output: {alias_output_data.outputId}")

alias_output = alias_output_data.output
updated_alias_output = wallet.get_client().build_alias_output(
    alias_id,
    unlock_conditions=alias_output.unlockConditions,
    state_index=int(alias_output.stateIndex) + 1,
    state_metadata=utf8_to_hex(NEW_STATE_METADATA),
    foundry_counter=alias_output.foundryCounter,
    immutable_features=alias_output.immutableFeatures,
    features=alias_output.features,
)

wallet.set_stronghold_password(os.environ["STRONGHOLD_PASSWORD"])

print('Sending transaction...')
transaction = account.send_outputs([updated_alias_output])
print(f'Transaction sent: {transaction.transactionId}')

# Wait for transaction to get included
blockId = account.retry_transaction_until_included(
    transaction.transactionId)
print(f'Block included: {os.environ["EXPLORER_URL"]}/block/{blockId}')
